### ISSUE TITLE
Liquibase prevents application startup if liquibase table is locked due to read operations requesting a lock

### ISSUE THREAD

**Issue by kequach** [liquibase/liquibase#6102]
### Search first

- [X] I searched and no similar issues were found

### Description

Hi, we are encountering an issue where our Spring Boot application running in Kubernetes was crashlooping after restarting due to Liquibase tables being locked during startup and Liquibase requesting a lock even for the read operations. 

We originally pinned the version to 4.20.0, since this was an issue that was reintroduced in 4.21.0 (see: https://github.com/liquibase/liquibase/issues/4312). This was fixed in version 4.23.1, which we currently have pinned: https://docs.liquibase.com/start/release-notes/liquibase-release-notes/liquibase-4.23.1.html
Relevant PR and issue: 
https://github.com/liquibase/liquibase/pull/4427
https://github.com/liquibase/liquibase/issues/4312

However, as mentioned, we encountered this exact issue even with the pinned version. Has anyone encountered similar issues and / or suggestions for a fix? Because of another open issue introduced in 4.24.0, we can not upgrade Liquibase yet: https://github.com/liquibase/liquibase/issues/5438

The actual stacktrace looks like this:

```
{
    "thread": "HikariPool-1 housekeeper",
    "logger": "com.zaxxer.hikari.pool.ProxyLeakTask",
    "loglevel": "WARN",
    "message": "Connection leak detection triggered for ConnectionID:1 ClientConnectionId: b6c919ac-d363-49ca-aec3-a3410bdd1c11 on thread main, stack trace follows",
    "stackTraces": [{
            "failureType": "java.lang.Exception",
            "failureMessage": "Apparent connection leak detected",
            "stackTrace": ["at com.zaxxer.hikari.HikariDataSource.getConnection(HikariDataSource.java:100)", "at liquibase.integration.spring.SpringLiquibase.afterPropertiesSet(SpringLiquibase.java:281)", "at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1833)", "at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1782)", "at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:600)", "at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:522)", "at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:325)", "at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)", "at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:323)", "at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:199)", "at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:312)", "at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:199)", "at org.springframework.context.support.AbstractApplicationContext.getBean(AbstractApplicationContext.java:1231)", "at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:949)", "at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:624)", "at org.springframework.boot.web.reactive.context.ReactiveWebServerApplicationContext.refresh(ReactiveWebServerApplicationContext.java:66)", "at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:754)", "at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:456)", "at org.springframework.boot.SpringApplication.run(SpringApplication.java:334)", "at org.springframework.boot.SpringApplication.run(SpringApplication.java:1354)", "at org.springframework.boot.SpringApplication.run(SpringApplication.java:1343)", "at com.bosch.bci.idm.solution.application.SolutionApplication.main(SolutionApplication.java:20)", "at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)", "at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)", "at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)", "at java.base/java.lang.reflect.Method.invoke(Method.java:568)", "at org.springframework.boot.loader.launch.Launcher.launch(Launcher.java:91)", "at org.springframework.boot.loader.launch.Launcher.launch(Launcher.java:53)", "at org.springframework.boot.loader.launch.JarLauncher.main(JarLauncher.java:58)"]
        }
    ]
}
```

### Steps To Reproduce

- Lock the db changelog table via DATABASECHANGELOGLOCK (set value from 0 to 1)
- Restart the application that runs Liquibase on startup without any new migrations

### Expected/Desired Behavior

Liquibase does not require a lock for the read operation

### Liquibase Version

4.23.1

### Database Vendor & Version

Azure SQL / Oracle

### Liquibase Integration

Spring Boot

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR (Thank you!)

**Comment by tati-qalified** [comment]
Hi @kequach, there was a [refactoring](https://github.com/liquibase/liquibase/pull/5925) for v4.29.1. Would you be able to upgrade to the latest version and see if it works?

Thank you,
Tatiana

**Comment by kequach** [comment]
Hi @tati-qalified thanks for your response! For some reason, I couldn't replicate this issue anymore. Instead, everytime I manually locked the table and restarted a service, the lock was ignored and removed.

**Comment by tati-qalified** [comment]
Good to hear! Feel free to reopen this issue if you come across the problem again.

**Comment by hubeni** [comment]
@tati-qalified 
Using Spring 3.5 which includes Liquibase v4.31.1, the issue still occurs: 
When the entry in the DATABASECHANGELOGLOCK table is locked and no migrations have to be done, the application is not able to start and remains in a inconsistent state.

As a double-check to the previous analysis, I've also pinned Liquibase to v4.29.1 which also reproduces the issue. 
So it does not look like that there has been a regression, but that the issue was never fixed.
Trying to understand better why the analysis of @kequach came to the conclusion that the issue was fixed in v4.29.1, I tried to reproduce the issue ~ 10 - 20 times. In one of the attempts, the lock seemed not to be an issue and the application started. But this looks rather like a concurrency topic / false positive of my manual testing than that it's sometimes a problem and sometimes not.

=> Could we re-open the ticket and investigate the issue?

**Comment by tati-qalified** [comment]
@kequach the issue appears to be that the way Spring works with Liquibase is that at application startup, Liquibase runs. This means that if Liquibase can't run successfully, startup won't be successful.

It doesn't seem to be an actual bug, but a different expectation for the current behaviour. This might be possible to implement (adding a configuration that allows to skip Liquibase runs if the DBCLL table is locked). Would you be willing to submit a PR to implement this change?

Thanks!
